beps/0005-definitive-plugin-listing/README.md

```diff
@@ -0,0 +1,250 @@
+  getInstalledFeatures: () => BackstageFeature[];
+}
+```
+
+Currently, this would need to be bootstrapped directly in `BackendInitializer` or `BackstageBackend`. It may make sense to create a new `FeatureRegistry` class that can expose this instead similar to the `PluginMetadataService`.
+
+### New `DeploymentMetadataService`
+
+This will be a new service that can be provided either as a core service or a new package. The idea is that it will show all information across your entire deployment. One can imagine this hosting aggregate information around instances in the deployment, like their internal and external URLs.
+
+```ts
+interface DeploymentMetadataService {
+  getInstalledFeatures: () => BackstageFeature[];
+}
+```
+
+By default, this will either
+a. be the exact same as `InstanceMetadataService` if we are not using `HostDiscovery` static config, or
+b. this will be pulled from that `HostDiscovery` static config.
+
+That would look something like,
+
+```yaml
+discovery:
+  endpoints:
+    - target: https://internal.example.com/internal-catalog
+      plugins: [catalog]
+    - target: https://internal.example.com/secure/api/{{pluginId}}
+      plugins: [auth, permission]
+    - target:
+        internal: https://internal.example.com/search
+        external: https://example.com/search
+      plugins: [search]
+```
+
+So, the above interface would return for the above config,
+
+```ts
+const installedFeatures = deploymentMetadataService.getInstalledFeatures();
+// installedFeatures = [{type: 'plugin', pluginId: 'catalog'}, {type: 'plugin', pluginId: 'auth'}, {type: 'plugin', pluginId: 'permission'}, {type: 'plugin', pluginId: 'search'}]
+```
+
+### New HTTP API to expose `InstanceMetadataService`
+
+We also propose exposing the new services through an HTTP API, so that these can be queried.
+
+```yaml
+paths:
+  /.backstage/instanceInfo/features/installed:
+    get:
+      summary: Get a list of installed features for this instance.
+      operationId: GetInstalledFeaturesByInstance
+      responses:
+        '200':
+          description: Successful operation
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  oneOf:
+                    - type: object
+                      properties:
+                        type:
+                          type: string
+                          enum:
+                            - plugin
+                        pluginId: { type: string }
+                    - type: object
+                      properties:
+                        type:
+                          type: string
+                          enum:
+                            - plugin
+                        pluginId: { type: string }
+                        moduleId: { type: string }
+```
+
+### New HTTP API to expose `DeploymentMetadataService`
+
+```yaml
+paths:
+  /.backstage/deploymentInfo/features/installed:
```

Comment:
So this is more for frontend consumption, while the other one is for service consumption - so to speak? As in, if we make a frontend plugin that wants to show an overview of your Backstage ecosystem, would it just pick some random plugin ID of sorts and call this enpoint there? And then that would perhaps use discovery and make a bunch of instance info calls to do its job?

Makes me wonder if the deployment info service shouldn't be a service, but rather a core plugin whose ID you could resolve with discovery and make calls to.

Reply:
Interesting point, I definitely agree - instance data needs to be a service to tie into startup information, but this could be spun off to a separate plugin for sure.